### PR TITLE
Add compiletime.ops.string.CharAt

### PR DESCRIPTION
**Pull request by OlivierBlanvillain**
This PR builds on #13400 to add a type-level `CharAt` operation. It's almost possible to define that function in user space:

```scala
type CharAt[R <: String, At <: Int] =
  Substring[R, At, At + 1]
```

The only difference being that `Substring` returns a `String`, while `CharAt` returns a `Char`.

@soronpo I've added you as a reviewer, let me know if you have time to have a look!

### CODE REVIEW

**Comment by OlivierBlanvillain** [comment]
> I'm not clear on motivation/use-case. Why is Char needed? And if it's needed why is just CharAt enough (and we don't need comparisons and other operations)?

In my mind length and charAt are the two string primitives that one needs to analyze strings. They give an array-like view that is rather universal, and work pretty well with our current support for type-level integers.

In terms of use-case, I've been working on regex library with type- and null-safe capture groups. It analyses string with recursive match types, and it's only missing `CharAt`. I haven't measured performance yet, but I suspect that working with character should be slightly better than building single character strings with `Substring`...

**Comment by soronpo** [comment]
OK. From the milestones I'm guessing that the compiler team is getting ready for 3.2.0-RC1. 
Since if this PR will only be added there, I think it is better to slightly expand the scope of it, drop the `@experimental` for all the compiletime operations and uncomment the `int.toString` [deprecation](https://github.com/lampepfl/dotty/blob/f08e8c05507db4ecc853df13e96a6226c50c30e2/library/src/scala/compiletime/ops/int.scala#L186). 
What do you think?

**Comment by OlivierBlanvillain** [comment]
I updated the annotations. Did I understand correctly that we don't need to mark the new operation as `@experimental` since the next release (3.2.0) is non-minor?

**Comment by smarter** [comment]
No, historically library additions didn't require SIP approval.

**Review by soronpo — APPROVED** [review]
LGTM

**Comment by smarter** [comment]
The baseVersion in the main branch is still 3.1.3-RC1, I don't know if it will be bumped before the next release so until this is changed we'd need the `@experimental` flag to merge this.

**Comment by soronpo** [comment]
Why wouldn't it be bumped before? AFAIK, there is no rush on this, so can wait for 3.2.0-RC1.

**Comment by smarter** [comment]
Well, we need a reason to bump, that can be because enough experimental changes have accumulated or because there's some big PR that requires a bump and we don't want to keep it open too long, I think that it'd make sense to do it now but it requires someone to take the time to go through experimental changes, etc.

**Review by Kordyjan — CHANGES_REQUESTED** [review]
All symbols that are made not-experimental should be marked with `@since("3.2")` annotation.

**Comment by OlivierBlanvillain** [comment]
Blocked on #14866 and its followups.

**Comment by mbovel** [comment]
Seems that we still cannot use `@since("3.2")` with `3.1.3-RC2` as the reference version.

Are these annotations still needed with the new versioning scheme? @Kordyjan

---

Are we sure we want to remove the experimental flags? @soronpo @OlivierBlanvillain 

I am asking because I am investigating more general ways to represent term applications at the type-level, which would make these types unneeded. For example, if we can represent the type of the result of an addition natively with something like `val z: {x + y} = x + y`, we would not need a separate `+` type anymore.

Also, there are still some open questions about the signatures of these types like singletonness bound and variance:

```scala
type +[A <: Singleton & Int, B <: Singleton & Int] <: Singleton & Int
type +[+A <: Int, +B <: Int] <: Int
type +[A <: Int, B <: Int] <: Int
```

(I am actually now pretty sure that I should revert https://github.com/lampepfl/dotty/pull/14452 and go with the third option.)

Is it a blocker for you to have these types as experimental @soronpo?

**Comment by soronpo** [comment]
> Are we sure we want to remove the experimental flags? @soronpo @OlivierBlanvillain

I think the feature is useful and was enough time under `@experimental`. There is no "surprise" element to it.

> Is it a blocker for you to have these types as experimental @soronpo?
Not a blocker for me. But maybe there are users of the singleton-ops library that it's a blocker for them to migrate to Scala 3.

> I am asking because I am investigating more general ways to represent term applications at the type-level, which would make these types unneeded. For example, if we can represent the type of the result of an addition natively with something like val z: {x + y} = x + y, we would not need a separate + type anymore.

It would be interesting to see. But this is a language change vs. `compiletime.ops` which is a library change. So I don't see this kind of this merged before Scala 3.3.0, so it may make sense to just merge this for users that require it and deprecate it later IF we find a better approach.

**Comment by Kordyjan** [comment]
> Are these annotations still needed with the new versioning scheme?

No. My previous comment is now outdated.

**Comment by mbovel** [comment]
We discussed this with the compiler team and decided that type-level operations should stay experimental for now, to let us time to evaluate potential alternatives.

**Comment by mbovel** [comment]
Oh, I didn't realize that some type-level operations were already stable.

So we re-discussed it and I got the approval for making all operations in this PR non-experimental 🥳

**Review by mbovel — APPROVED** [review]

**Comment by mbovel** [comment]
Let's get that merged :)

**Comment by OlivierBlanvillain** [comment]
If that's OK I'll let someone else pull the trigger because of the release/version implications, which I'm not really on top of :)

File: tests/run-custom-args/tasty-inspector/stdlibExperimentalDefinitions.scala
Comment (nicolasstucki) [inline tests/run-custom-args/tasty-inspector/stdlibExperimentalDefinitions.scala:50]: @mbovel didn't we say that these operations would stay experimental?

Comment (mbovel) [reply tests/run-custom-args/tasty-inspector/stdlibExperimentalDefinitions.scala:50]: See https://github.com/lampepfl/dotty/pull/14431#issuecomment-1129030756; when we first discussed it, I hadn't realized that some compile-time operations were already stable. Given this, I re-discussed it with Martin and he agreed to make these operations stable as well.

The argument is that operations on `long`, `float` and `double` are exactly the same as the ones on `int` which are already stable, so it makes sense to make them stable as well. In term of maintenance, this doesn't change much.

Including operations on strings as well makes sense are they are important building blocks to do anything useful with strings at the type-level (Olivier used them for his work on regular expressions for example).

**Pull request merged**
